mudlib: small bugfixes for query_ip_name and query_ip_number simuls.

cpu opened the pull request:
:wave: Hi folks,

Thanks again for providing the `mudlib/deprecated/` simuls. They were extremely helpful to me.

Here are two small commits that update the `query_ip_number` and `query_ip_name` simuls based on my experience migrating from 3.2.17 to 3.5.4. 

### mudlib: guard query_ip_x simuls for interactives - 84beba5

The deprecated simul backfills for `query_ip_name` and `query_ip_number` exhibit a slight difference in behaviour compared to how the efun counterparts appeared to have worked on 3.2.x.

When given a linkdead non-interactive player object the simuls throw an error where previously the efuns would return 0.

On 3.2.17:
```
Guest (P:1:none) goes linkdead.
lpc return query_ip_name(find_player("guest"));
[55|0.000012s] Result: 0
```

On 3.5.4:
```
Guest (P:1:none) goes linkdead.
lpc return query_ip_name(find_player("guest"));
RUNTIME ERROR (CAUGHT): /secure/simul_efun (/secure/simul/compat/query_ip_name.c) line 16: Bad arg 1 to interactive_info(): Object 'obj/player/player#985' is not interactive.
```

This commit introduces a fix to both simuls by checking if `ob` is `interactive()` before calling `efun::interactive_info` with `ob` as an arg. If a non-interactive `ob` is provided both simuls now return `0`.

### mudlib: fix type of query_ip_x simul player arg. -  e87f0e9

Per discussion with @zesstra the correct type for the `player` argument passed by reference to `query_ip_name` and `query_ip_number` should be `object|int*` to support the legacy behaviour as well as the `interactive_info(ob, II_IP_ADDRESS) ` assignment.

cpu commented:
> Rather than return 0 for non-interactive calls we return "" as this seems like a better default in this case.

My LPC is rusty :-) I forgot `0` is a valid return for `string`. Probably best to use that. I will push a revision shortly.

cpu commented:
> My LPC is rusty :-) I forgot 0 is a valid return for string. Probably best to use that. I will push a revision shortly.

Fixed with force-push. Updated PR description accordingly.

zesstra commented:
The reason for the assignment to player (at least in query_ip_number) is this:

        If ob is given as reference (and it must be a valid object then),
        it will upon return be set to the struct sockaddr_in of the queried
        object, represented by an array of integers, one integer per
        address byte:
          ob[0.. 1]: sin_family
          ob[2.. 3]: sin_port
          ob[4.. 7]: sin_addr
          ob[8..15]: undefined.

But you are right, that type is incorrect as-is.
Also, I have to check the real behaviour of the old query_ip_name efun yet.

cpu commented:
@zesstra Thanks for taking a look!

> But you are right, that type is incorrect as-is.

Beyond the type it looks like the result with the `sockaddr_in` bytes isn't used for anything, right? :-) 

Like we could switch it to be `bytes sockaddr_in = efun::interactive_info(ob, II_IP_ADDRESS);` but I'm still not seeing a reason why that information is required in the simul and when I removed it entirely it appeared to work OK.

zesstra commented:
You can't know, what the caller does with the information or does not.
If <player> is given by-reference, the assignment changes the variable the caller passes. And the caller can then work with the information after the call.
So, basically removing the assignment will break any usage where the caller passes <player> by reference.

Nowadays, interactive_info(ob, II_IP_ADDRESS) returns an array of bytes.
But query_ip_number() did return an array of ints and the replacement sefun should do as well, because otherwise the (historic) callers of the sefun will get errors.

So, I believe, the correct type for <player> should be: <object|int*>

cpu commented:
> If 'player' is given by-reference, the assignment changes the variable the caller passes. And the caller can then work with the information after the call.

Ohhhhhhhhh! That's the intuition I was missing. Yes that makes perfect sense now. Thanks for the explanation. 

> So, basically removing the assignment will break any usage where the caller passes 'player' by reference.

Aha, and I suspect the lib i'm testing with just happened to have none. Luck not correctness :-)

> So, I believe, the correct type for 'player' should be: <object|int*>

Ok! I will rework 5e7cb18 later this evening to correct the cast instead of dropping the statement. Thanks @zesstra !

cpu commented:
> Ok! I will rework 5e7cb18 later this evening to correct the cast instead of dropping the statement.

Reworked with e87f0e9 and updated the PR description. Hopefully I understood what you were recommending. Let me know if I missed the mark and I will continue to iterate!

amotzkau commented:
Merged.

cpu commented:
Thanks :-)

amotzkau commented:
Au contraire, we thank you for your contributions. :-)
